core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -161,42 +184,22 @@
       // invalid scheme. Just fall through and will try "dns:///foo.googleapis.com:8080"
       uriSyntaxErrors.append(e.getMessage());
     }
-    if (nameResolver == null) {
+    // It may be Opaque if the host was parsed as the scheme
+    if (targetUri == null || targetUri.isOpaque()) {
```

Comment:
Although `isOpaque()` works for the `localhost:80` case for now, I think this is abusing `isOpaque()`, and it adds a restriction that the scheme-specific part MUST start with "/" to make it non-opaque. I still prefer the current logic, that we try `newNameResolver()` with the original URI first, and if it returns `null`, try with `dns:///` again.
@ejona86 WDYT?